Authenticated Event Streams

alex-mysten opened the pull request:
## Description 

This PR implements "Authenticated Event Streams". Users can emit events via emit_authenticated() in move smart contracts which are accumulated into a Merkle Mountain Range stored on the checkpoint, per-stream.

## Test plan 

UTs, transactional test, move tests, and e2e test

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] gRPC:
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

mystenmark commented on sui-execution/latest/sui-adapter/src/programmable_transactions/context.rs:
```diff
@@ -1547,6 +1547,23 @@ mod checked {
+                        address,
+                        operation: accum_event.action.into_sui_accumulator_action(),
+                        value,
+                    };
```
seems this can be factored by having one `map()` that maps `MoveAccumulatorValue` to `AccumulatorValue`, and then the rest of the code can be in a second `map()` that handles both cases

alex-mysten replied:
Done

mystenmark commented on crates/sui-protocol-config/src/lib.rs:
```diff
@@ -3964,6 +3966,9 @@ impl ProtocolConfig {
                     // Reudce the frequency of checkpoint splitting under high TPS.
                     cfg.max_transactions_per_checkpoint = Some(20_000);
                 }
+                96 => {
```
FYI v95 is still not in any release branch, although maybe by the time this lands it will be

alex-mysten commented on crates/sui-framework/packages/sui-framework/sources/event.move:
```diff
@@ -37,6 +45,101 @@ module sui::event;
+    let mut i = 0;
+    let mut cur = new_val;
+
+    while (i < vector::length(mmr)) {
```
Unless I'm missing something, I think this implementation is correct. We are essentially doing a binary increment with carry with mmr[0] as the LSB. We only push_back when every `mmr[i]` was set and we're hashing all of them together to create a new MSB.

alex-mysten commented on crates/sui-core/src/accumulators/mod.rs:
```diff
@@ -107,10 +130,34 @@ impl From<MergedValueIntermediate> for MergedValue {
+}
+
+fn build_event_merkle_root(events: &[EventCommitment]) -> Digest {
+    if events.is_empty() {
```
Good idea - updated :)

alex-mysten commented on crates/sui-framework/packages/sui-framework/sources/accumulator_settlement.move:
```diff
@@ -68,3 +70,147 @@ fun settle_u128<T>(
 native fun record_settlement_sui_conservation(input_sui: u64, output_sui: u64);
+
+#[allow(unused_field)]
+public struct EventStreamHead has store {
```
Correct

alex-mysten replied:
yes, one per stream, sorry

tzakian commented on crates/sui-protocol-config/src/lib.rs:
```diff
@@ -263,6 +263,7 @@ const MAX_PROTOCOL_VERSION: u64 = 96;
 // Version 96: Enable authority capabilities v2.
 //             Fix bug where MFP transaction shared inputs' debts were not loaded
 //             Create Coin Registry object
+// Version 97: Add constant for event emit auth stream cost.
```
Nit: You also added authentication stream support to the sui-framework and sui-framework upgrades are (implicitly) a part of the protocol versions so we try to include that here as well (e.g., just a simple "Added support to sui-framework for authenticated event streams" is fine).

tzakian reviewed (commented):
Overall looks good.

Mainly small things in the comments. Three important things:
1. Just want to double-check the protocol gating in the new native as this may need to change if we are not rolling this out to `mainnet` immediately (left a comment on this). It looks like we are from the protocol config, but just want to double-check on that.
2. If you could add a test in this file: https://github.com/MystenLabs/sui/blob/main/crates/sui-verifier-transactional-tests/tests/private_generics/private_event_emit.mvir (or recreate this test e.g., in `private_event_authenticated_emit.mvir`) and make sure the private generic rules fire correctly in the sui verifier.
3. We need to update the compiler to support checking the private entry rules around `authenticated_emit` and error on them here https://github.com/MystenLabs/sui/blob/main/external-crates/move/crates/move-compiler/src/sui_mode/typing.rs#L975 and add tests for it here basically copying this test and `s/emit/authenticated_emit` https://github.com/MystenLabs/sui/blob/main/external-crates/move/crates/move-compiler/tests/sui_mode/private_generics/private_event_emit.move. If you want I can put up a PR against this branch adding it in too -- it shouldn't take me too long. But let me know what is preferable for you (if you'd prefer to do it, or if you'd prefer I do it no problem at all just ping me :) )

alex-mysten commented:
> Overall looks good.
> 
> Mainly small things in the comments. Three important things:
> 
> 1. Just want to double-check the protocol gating in the new native as this may need to change if we are not rolling this out to `mainnet` immediately (left a comment on this). It looks like we are from the protocol config, but just want to double-check on that.
> 2. If you could add a test in this file: https://github.com/MystenLabs/sui/blob/main/crates/sui-verifier-transactional-tests/tests/private_generics/private_event_emit.mvir (or recreate this test e.g., in `private_event_authenticated_emit.mvir`) and make sure the private generic rules fire correctly in the sui verifier.
> 3. We need to update the compiler to support checking the private entry rules around `authenticated_emit` and error on them here https://github.com/MystenLabs/sui/blob/main/external-crates/move/crates/move-compiler/src/sui_mode/typing.rs#L975 and add tests for it here basically copying this test and `s/emit/authenticated_emit` https://github.com/MystenLabs/sui/blob/main/external-crates/move/crates/move-compiler/tests/sui_mode/private_generics/private_event_emit.move. If you want I can put up a PR against this branch adding it in too -- it shouldn't take me too long. But let me know what is preferable for you (if you'd prefer to do it, or if you'd prefer I do it no problem at all just ping me :) )

@tzakian 

1. My plan was to roll this out immediately. I updated the comment to reflect that intention
2. Added tests in private_event_emit
3. Updated and tests added

Thanks for the guidance!

alex-mysten commented on sui-execution/latest/sui-move-natives/src/event.rs:
```diff
@@ -41,15 +47,61 @@ pub fn emit(
+    context: &mut NativeContext,
+    mut ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
```
I did not plan to gate usage of this via protocol config

jessiemongeon1 reviewed (commented):
Docs changes are autogenerated and LGTM

mystenmark commented on crates/sui-core/src/accumulators/mod.rs:
```diff
@@ -107,10 +130,34 @@ impl From<MergedValueIntermediate> for MergedValue {
+    digest: Digest,
+}
+
+fn build_event_merkle_root(events: &[EventCommitment]) -> Digest {
```
reminder to add debug check that events is sorted (it should be by default)

tzakian reviewed (approved):
Looks good to me! ✅
